go-controller/pkg/ovn/egressip.go

```diff
@@ -2510,52 +2015,80 @@ func (e *egressIPController) deleteReroutePolicyOps(ops []ovsdb.Operation, podIP
+		nextHopIP, err = e.getTransitSwitchIP(status.Node)
+		if apierrors.IsNotFound(err) {
+			// if the node is already gone, then don't count it as error as during deletion.
+			klog.Warningf("Unable to fetch transit switch IP for node: %s: err: %v", status.Node, err)
+		} else if err != nil {
+			return nil, fmt.Errorf("unable to fetch transit switch IP for node %s: %v", status.Node, err)
+		}
 	}
 
 	var ops []ovsdb.Operation
-	if gatewayRouterIP != nil {
-		gwIP := gatewayRouterIP.String()
+	if nextHopIP != "" {
 		policyPred := func(item *nbdb.LogicalRouterPolicy) bool {
 			hasGatewayRouterIPNexthop := false
 			for _, nexthop := range item.Nexthops {
-				if nexthop == gwIP {
+				if nexthop == nextHopIP {
 					hasGatewayRouterIPNexthop = true
 					break
 				}
 			}
 			return item.Priority == types.EgressIPReroutePriority && item.ExternalIDs["name"] == name && hasGatewayRouterIPNexthop
 		}
-		ops, err = libovsdbops.DeleteNextHopFromLogicalRouterPoliciesWithPredicateOps(e.nbClient, nil, types.OVNClusterRouter, policyPred, gwIP)
+		ops, err = libovsdbops.DeleteNextHopFromLogicalRouterPoliciesWithPredicateOps(e.nbClient, ops, types.OVNClusterRouter, policyPred, nextHopIP)
 		if err != nil {
 			return nil, fmt.Errorf("error removing nexthop IP %s from egress ip %s policies on router %s: %v",
-				gatewayRouterIP, name, types.OVNClusterRouter, err)
+				nextHopIP, name, types.OVNClusterRouter, err)
 		}
 	}
 
-	routerName := util.GetGatewayRouterFromNode(status.Node)
-	natPred := func(nat *nbdb.NAT) bool {
-		return nat.ExternalIDs["name"] == name && nat.ExternalIP == status.EgressIP
-	}
-	nats, err := libovsdbops.FindNATsWithPredicate(e.nbClient, natPred) // save the nats to get the podIPs before that nats get deleted
-	if err != nil {
-		return nil, fmt.Errorf("error removing egress ip pods from adress set %s: %v", EgressIPServedPodsAddrSetName, err)
-	}
-	ops, err = libovsdbops.DeleteNATsWithPredicateOps(e.nbClient, ops, natPred)
-	if err != nil {
-		return nil, fmt.Errorf("error removing egress ip %s nats on router %s: %v", name, routerName, err)
+	var nats []*nbdb.NAT
+	if !config.OVNKubernetesFeature.EnableInterconnect || (ok && isLocalZoneNode.(bool)) {
+		if nextHopIP != "" {
+			p := func(item *nbdb.LogicalRouterStaticRoute) bool {
+				return item.Nexthop == nextHopIP && item.ExternalIDs["name"] == name && item.Policy != nil && *item.Policy == nbdb.LogicalRouterStaticRoutePolicySrcIP
+			}
+			ops, err = libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicateOps(e.nbClient, ops, types.OVNClusterRouter, p)
+			if err != nil {
+				return nil, fmt.Errorf("error deleting logical router static routes on router %s for %s: %v", types.OVNClusterRouter, name, err)
+			}
+		}
+		routerName := util.GetGatewayRouterFromNode(status.Node)
+		natPred := func(nat *nbdb.NAT) bool {
+			// TODO(tssurya): investigate if we should be deleting NATs only from status.Node
+			return nat.ExternalIDs["name"] == name && nat.ExternalIP == status.EgressIP
```

Comment:
oh, very good point. In this new world, we may be more susceptible to have mid air collisions where node X is removing and node Y is adding the nat for the given eip due to a revalidation. But since they are distinct NB dbs, it may be okay, agree?

Reply:
yes +1 in IC world we should be good since DBs will be distinct.